### PR TITLE
fix(users): add password validations

### PR DESCRIPTION
**Pull request by Riddhiagrawal001**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
The mandatory condition for the password field was just not be empty . Following the PR, the password field will be subject to the following checks:
1.The password must be between 8 and 70 characters in length.
2.It must include at least one uppercase character.
3.It must include at least one lowercase character.
4.It must include at least one special character.
5.It must include at least one numeric character.
6.It must not contain whitespace.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Closes https://github.com/juspay/hyperswitch/issues/4412

## How did you test it?
1.Signup API ( local testing)
```
curl --location 'http://localhost:8080/user/signup' \
--header 'Content-Type: application/json' \
--data-raw '{
    "email": "email value",
    "password": "456667sdvh",
    "country": "IN"
}'
```
Failure case response
``` {
    "error": {
        "type": "invalid_request",
        "message": "Invalid Password",
        "code": "UR_09"
    }
} 
```

2.Reset Password API (local testing)
```
curl --location 'http://localhost:8080/user/reset_password' \
--header 'Content-Type: application/json' \
--data '{
    "password": "456667sdvd",
    "token":"token from email after clicking Reset password in dashboard"
}'
```
Failure case response
``` 
{
    "error": {
        "type": "invalid_request",
        "message": "Invalid Password",
        "code": "UR_09"
    }
} 
```

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/utils/user/password.rs [hunk 1]
```diff
@@ -38,3 +39,20 @@ pub fn is_correct_password(
     }
     .change_context(UserErrors::InternalServerError)
 }
+
+pub fn get_temp_password() -> String {
+    let uuid_pass = uuid::Uuid::new_v4().to_string();
+    let mut rng = rand::thread_rng();
+
+    let special_chars: Vec<char> = "!@#$%^&*()-_=+[]{}|;:,.<>?".chars().collect();
+    let specialchars: char = *special_chars.choose(&mut rng).unwrap_or(&'@');
```
Comment (apoorvdixit88) [inline crates/router/src/utils/user/password.rs:48]: variable name `special_char`

File: crates/router/src/types/domain/user.rs [hunk 2]
```diff
@@ -158,6 +158,40 @@ pub struct UserPassword(Secret<String>);
...
+        let is_too_short = password.graphemes(true).count() < consts::user::MIN_PASSWORD_LENGTH;
+
+        if is_too_short || is_too_long || !is_password_format_valid {
+            Err(UserErrors::PasswordParsingError.into())
+        } else {
+            Ok(Self(password.into()))
+        }
+    }
+
+    pub fn new_without_validation(password: Secret<String>) -> UserResult<Self> {
```
Comment (apoorvdixit88) [inline crates/router/src/types/domain/user.rs:193]: may be `new_password_without_validation`,
we are not validating the password in merchant create request?

**Review by apoorvdixit88 — COMMENTED** [review]
`The password must be between 8 and 50 characters in length`
Please change description, as per code the maximum length allowed is 70

Also if possible swap description and motivation and context content.

File: crates/router/src/types/domain/user.rs (hunk 2 above)
Comment (Riddhiagrawal001) [reply crates/router/src/types/domain/user.rs:193]: We're retrieving user details from the database to create a new merchant. It's possible that some old users didn't followed the password convention. Therefore, we've implemented a workaround specifically for this scenario to bypass validation.

File: crates/router/src/utils/user/password.rs (hunk 1 above)
Comment (Riddhiagrawal001) [reply crates/router/src/utils/user/password.rs:48]: Changed to `specialchars` to special_char

**Review by apoorvdixit88 — APPROVED** [review]

**Review by racnan — CHANGES_REQUESTED** [review]

File: crates/router/src/utils/user/password.rs [hunk 3]
```diff
@@ -38,3 +39,20 @@ pub fn is_correct_password(
     }
     .change_context(UserErrors::InternalServerError)
 }
+
+pub fn get_temp_password() -> String {
+    let uuid_pass = uuid::Uuid::new_v4().to_string();
+    let mut rng = rand::thread_rng();
+
+    let special_chars: Vec<char> = "!@#$%^&*()-_=+[]{}|;:,.<>?".chars().collect();
+    let special_char: char = *special_chars.choose(&mut rng).unwrap_or(&'@');
```
[the remark this answers is not in the document]
Comment (Riddhiagrawal001) [reply crates/router/src/utils/user/password.rs]: Addressed

**Review by racnan — DISMISSED** [review]
LGTM

**Review by ThisIsMani — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Pull request merged**